File: hdu/修复公路.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

void dfs(const vector<vector<int>>& e, vector<bool>& tag, int i) {
    stack<int> stk;
    stk.push(i);
    
    while (!stk.empty()) {
        int current = stk.top();
        stk.pop();
        
        if (tag[current]) {
            continue;
        }
        
        tag[current] = true;
        
        for (int j : e[current]) {
            if (!tag[j]) {
                stk.push(j);
            }
        }
    }
}

int main() {
    std::ios::sync_with_stdio(0);
  std::cin.tie(0);
    int t;
    cin >> t;
    
    while (t--) {
        int n;
        cin >> n;
        
        vector<int> a(n);
        for (int i = 0; i < n; ++i) {
            cin >> a[i];
        }
        
        vector<vector<int>> e(n);
        for (int i = 0; i < n; ++i) {
            if (i + a[i] < n) {
                e[i].push_back(i + a[i]);
                e[i + a[i]].push_back(i);
            }
            if (i - a[i] >= 0) {
                e[i].push_back(i - a[i]);
                e[i - a[i]].push_back(i);
            }
        }
        
        vector<bool> tag(n, false);
        int cnt = 0;
        
        for (int i = 0; i < n; ++i) {
            if (!tag[i]) {
                dfs(e, tag, i);
                cnt++;
            }
        }
        
        cout << cnt - 1 << endl;
    }
    
    return 0;
}

File: hdu/修复公路.py
from collections import defaultdict


def dfs(e, tag, i):
    if tag[i]:
        return
    else:
        tag[i] = True
    for j in e[i]:
        dfs(e, tag, j)
    # if i-r>=1:
    #     dfs(lst,tag,i-r)
    # if i+r<len(lst):
    #     dfs(lst,tag,i+r)


for _ in range(int(input())):
    n = int(input())
    a = map(int, input().split())
    e = [list() for _ in range(n)]
    for i, v in enumerate(a):
        if i + v < n:
            e[i].append(i + v)
            e[i + v].append(i)
        if i - v > 0:
            e[i].append(i - v)
            e[i - v].append(i)
    tag = [False] * (n)
    cnt = 0
    for i in range(0, n):
        if not tag[i]:
            dfs(e, tag, i)
            cnt += 1
    print(cnt - 1)
